### PR TITLE
Holder bindings

### LINKED ISSUE #12: Subject Holder restriction
In the case of VC responses, it looks to be assumed that the subject and holder are the same identity. 
Is an explicit selector needed for this (or an explicit exception) e.g. You may hold a vaccination credential for a dog which is the subject of said credential. Or is it recommended that this be handled in business logic.

### LINKED ISSUE #28: Demonstrate proof of link secret constraints for fields
In reference to #24 

We must support the specification of a proof of link secret constraint as expressed as a base64 encoded data field. The contents are not necessarily known ex-ante in their entirety or sometimes at all. Instead, implementation-specific algorithms must be ran to ensure their integrity and validity.

### LINKED ISSUE #47: how to allow verifiers to require multiple credential inputs to be about same subject
As raised in #18 

>    // TODO: how do you require that banking input 1 and banking
>    // input 2 both have the same credential subject (e.g., same
>    // primary account holder)?

If values from two different credentials are requested, the verifier should be able to specify that the attributes must have the same credential subject, or to specify the credential subject for each requested attribute.

### PR DESCRIPTION
**Pull request by brentzundel**
This PR adds a description of various holder bindings to the spec.

More discussion is needed to explore ways to incorporate these holder bindings into the submission requirements or input descriptors of a presentation definition.

fixes #28, fixes #12 

Signed-off-by: Brent Zundel <brent.zundel@gmail.com>

### CODE REVIEW

File: spec/spec.md [hunk 1]
```diff
@@ -1268,6 +1268,54 @@ information to resolve the status of a credential.
...
+Some credential signatures support the inclusion of holder-provided information
+that becomes incorporated into the signature, but remains hidden from the
+credential issuer. One common use of this capability is known as a link secret.
+A `link secret` is a value known only to the credential holder that is embedded
+in the signature of each issued credential. Upon presentation to a verifier, the
+holder demonstrates knowledge of the random value without revealing it, and
+proves that the same value is contained in every credential included in the
+proof presentation. The verifier can verify that the presenter of the
+credentials knows the link secret, and that the `link secret` is the same in
+each credential.
```
Comment (glcohen) [inline spec/spec.md:1303]: wary of putting impl specific ideas into this spec

Comment (brentzundel) [reply spec/spec.md:1303]: This section is intended to be non-normative and informational. It does touch on some specific aspects of how link secret proofs work, but I tried to make it general enough that any link secret implementation would roughly follow the same steps (much like the proof of control section talks specifically about demonstrating control through proving knowledge of the private key, but is general enough that most schemes would be applicable).
Could you recommend a re-wording that would generalize it enough to be acceptable?

In general, I'm hoping we can settle on a mechanism (ideally one that makes use of submission requirements or input descriptors) that could be used generally by a verifier to indicate its requirements around holder binding. Arguably, we've gotten part of the way there with the `subject_is_holder` flag, but that doesn't cover the cases where 1) a subject gives consent for the presentation (but isn't the holder), or 2) the verifier wants proof that multiple credentials are about the same subject.
I'm also wondering if holder binding is the most accurate term. Or perhaps it's fine, but I'm trying to address more than that here, which would indicate that another term (and another PR) may be required.

Mostly this was my first stab at trying to wrangle a couple of complicated use cases/requirements, so I appreciate the feedback!

Comment (glcohen) [reply spec/spec.md:1303]: I have no problem with it, just a call out. I think the approach itself is generic, the addition is great.

**Comment by glcohen** [comment]
hqdefault

**Comment by brentzundel** [comment]
Per our conversation during a presentation exchange spec drafting meeting, I have modified this PR to change the `subject_is_holder` property such that it can specify which claim subject is required to be the holder, in the case of multiple claim subjects.
Bikeshedding of property names welcome.

**Comment by dhh1128** [comment]
I am happy with the intent of the latest change, but I don't think the mechanism works. The reason is that I think the `type` property of a credential subject is unlikely to be specific enough. The example shows `type = "mother"`, but I think a far more likely case would be `type="person"`, which isn't going to distinguish the different subjects of a birth certificate.

I think we need something where we can declare a "role" for a subject and specify the criterion by role (prove something about the credential subject having a role = father, and prove something else about the credential subject having role = mother -- instead of proving something about the credential subject having type = parent or type = person).

**Comment by brentzundel** [comment]
@glcohen @csuwildcat 
I've modified the PR so it reflects the changes we discussed in the meeting today, but I think there may be a problem.
I changes the `is_holder` object so it would look like this:
```
"is_holder": {
	"id1": "required",
	"id2": "preferred"
}
```
but when I went to update the JSON Schema, I couldn't figure out how to do it. How do you define a property name (i.e., "id1") such that it needs to be the value of another property (i.e., the value of the `id` property in a `fields` object).

looking for help/tips/feedback.

**Comment by OR13** [comment]
possible solution? https://stackoverflow.com/questions/40741989/json-schema-to-validate-objects-values-against-content-of-another-object

**Comment by brentzundel** [comment]
@glcohen @csuwildcat @OR13 @wyc this PR is updated according to the meeting today and ready for review.

File: spec/spec.md [hunk 2]
```diff
@@ -1026,21 +1026,37 @@ Descriptor Objects_ are composed as follows:
...
+            value in the `field_id` array is the same as the entity submitting
+            the response.
+          - `preferred` - This indicates that it is ****RECOMMENDED**** that the
+            processing entity include proof that the subject of each attribute
+            identified by a value in the `field_id` array is the same as the
+            entity submitting the response.
+                                          
+        The `is_holder` property would be used by a [[ref:Verifier]] to
+        require that certain inputs be provided by a certain subject. For
+        example, an identity verification `presentation definition` might
```
Comment (glcohen) [inline spec/spec.md:1051]: nit `presentation_definition` --> `[[ref:Presentation Definition]]`

**Review by glcohen — APPROVED** [review]
looks good and fine to merge. one comment is perhaps it would be beneficial to add an example, or update an existing one that uses these properties

**Pull request merged**